grapevine/src/subscriber/aggregator.rs

```diff
@@ -0,0 +1,205 @@
+                            break;
+                        }
+                        empty = true;
+                    },
+                    Err(TryRecvError::Disconnected) => {
+                        break;
+                    },
+                }
+            }
+        });
+
+        Self {
+            state_rx,
+        }
+    }
+
+    /// subscribes to recent changes of state
+    pub fn subscribe(&self) -> watch::Receiver<State<S>> {
+        self.state_rx.clone()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+    use async_stream::stream;
+    use tokio::net::UdpSocket;
+    use tokio_stream::wrappers::WatchStream;
+    use futures::Stream;
+    use tokio::time::timeout;
+    use tokio_stream::StreamExt;
+    use crate::subscriber::message::test_utils::make_raw_msg_bytes;
+    use crate::subscriber::sequencer::tests::{TestStateSync, make_default_sequencer};
+
+    use super::*;
+
+    fn prefix_sum(arr: Vec<i64>) -> Vec<i64> {
+        let mut ps = Vec::with_capacity(arr.len());
+        for v in arr {
+            if let Some(last) = ps.last() {
+                ps.push(last + v);
+            } else {
+                ps.push(v);
+            }
+        }
+        ps
+    }
+
+    async fn publish_chunks(sizes: Vec<i64>, publisher: UdpSocket) {
+        let wait_between_chunks = Duration::from_millis(10);
+
+        tokio::spawn(async move {
+            let stream = stream! {
+                let mut seq_id = 0;
+                for size in sizes {
+                    for _ in 0..size {
+                        yield seq_id;
+                        seq_id += 1;
+                    }
+                    tokio::time::sleep(wait_between_chunks).await;
+                }
+            };
+            publish_to_socket(stream, publisher).await?;
+
+            Ok::<_, anyhow::Error>(())
+        });
+    }
+
+    async fn publish_to_socket(
+        stream: impl Stream<Item = i64>,
+        publisher: UdpSocket,
+    ) -> anyhow::Result<()> {
+        let mut stream = Box::pin(stream);
+        while let Some(i) = stream.next().await {
+            publisher.send(make_raw_msg_bytes(i)?.as_ref()).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn collect_result(
+        aggregator: Aggregator<TestStateSync>,
+        expected_len: usize
+    ) -> anyhow::Result<Vec<i64>> {
+        let result_stream = WatchStream::from_changes(aggregator.subscribe())
+            .take(expected_len)
+            .then(|v| async move { v.read().await.sum })
+            .collect::<Vec<_>>();
+
+        let result = timeout(Duration::from_secs(5), result_stream).await?;
+
+        Ok(result)
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn chunks_are_valid() -> anyhow::Result<()> {
+        // too big chunks can be divided (especially in multi-threaded runtime) which is decreasing test reproducibility
+        let sizes = vec![3, 5, 10, 3, 2, 2, 1, 1, 1];
```

Comment:
From how I'm interpreting this I can't see how the multi-threaded runtime is affecting reproducability. Could you add more context as to what the exact problem is? e.g. are packets getting dropped?

What goes wrong with larger chunks? Seeing as there's a sequencer and they're all sequenced in sync feels like should be fine.

Reply:
added examples. packets are not dropped. when chunks is becoming big, aggregator is starting to split them early (because channel says that it is drained, but in fact producer is still sending messages for this chunk)